[SE-0288] [stdlib] Adding isPower(of:) to BinaryInteger

dingobye opened the pull request:
This PR presents the implementation of adding a public API `isPower(of:)` to the `BinaryInteger` protocol.

- [Pitch Thread](https://forums.swift.org/t/adding-ispowerof2-to-binaryinteger/24087)
- Proposal: [SE-0288](https://github.com/apple/swift-evolution/blob/master/proposals/0288-binaryinteger-ispower.md)

dingobye committed:
Add 'isPower(of:)' to the 'BinaryInteger' protocol.

This PR adds a public API 'isPower(of:)', as an extension method, to the
'BinaryInteger' protocol.  It checks if an integer is power of a given base.

Resolves: SE-0288

dingobye commented:
The idea of this implementation is based on a fast-/slow-path pattern. We have a generic implementation, which is suitable for all inputs, as the slow path; and meanwhile provide some particularly optimized implementation for frequently used inputs (e.g., 2 and 10) as the fast paths.

Some experiments (code can be found [here](https://github.com/dingobye/swift/blob/is_power_of_performance/performance.swift)) are conducted to compare the performance of different scenarios, including:
- `isPower(of:)` the public API. It firstly tests if the input base equals 2 or 10. If so, it goes the fast path by calling the corresponding optimized version (i.e., `_isPowerOfTwo` or `_isPowerOfTen`); otherwise, it falls into the slow path (i.e., `_isPower(of:)`).
- `_isPowerOfTwo` the fast path for input base 2, with the classic implementation `n > 0 && n & (n - 1) == 0`
- `_isPowerOf2_ctpop` an alternative fast path for input base 2 with the popcount implementation
- `_isPowerOfTen` the fast path for input base 10, taken from [Michel Fortin's comments](https://forums.swift.org/t/adding-ispowerof2-to-binaryinteger/24087/38)
- `_isPower(of:)` the general implementation (i.e., the slow path) for any input

</br>

The table below shows the execution time (in sec) for type `Int64` under the `-O` optimization level across a range of platforms. The results for other built-in integer types have a similar pattern to this one, so they are not given and discussed here.

. | MacBook Pro (2.9G Core i7) |  iPhone X Max | iPhone 6S 
-------- | -------- | -------- | -------- 
isPower(of: 2) | 0.112 |  0.121 | 0.233
_isPowerOfTwo |  0.113 | 0.121 | 0.233
_isPowerOfTwo_ctpop | 0.178 |  0.121 | 0.268
_isPower(of: 2) | 2.184 | 4.757 | 9.128
_isPower(of: 4) | 1.097 | 1.360 | 4.282
_isPowerOfTen | 0.257 |  0.201 | 0.333
_isPower(of: 10) | 0.739 |  0.995 | 2.663

</br>

The table below illustrates the results for [BigInt](https://github.com/apple/swift/blob/master/test/Prototypes/BigInt.swift) (with 1024-bit width). The configuration is the same as for built-in integers, except that the workload is repeated in a much less amount.

. | MacBook Pro (2.9G Core i7) | iPhone X Max | iPhone 6S 
-------- | -------- | -------- | -------- 
isPower(of: 2) | 0.004 | 0.003 | 0.008
_isPowerOfTwo |  0.003 | 0.003 | 0.007
_isPower(of: 2) | 1.076 | 1.164 | 3.066
_isPower(of: 4) | 0.822 | 0.750 | 1.975
_isPowerOfTen | 0.644 | 0.475 | 1.232
_isPower(of:10) | 0.928 | 0.683 | 1.770

</br>

It can be seen that:
- `isPower(of: 2)` (i.e., calling the public interface) has similar performance as the fast path `_isPowerOfTwo` for built-in integers, and slightly less efficient for BigInt.
- In addition, the popcount implementation can be as efficient as `_isPowerOfTwo` on some latest iphone device (e.g., iPhone X Max), but overall is less performant. Another problem for popcount implementation is it narrows down the scope from `BinaryInteger` to `FixedWidthInteger`.
- Moreover, for both particularly optimized cases (i.e., `_isPowerOfTwo` and `_isPowerOfTen`), they are significantly more performant than the generic implementation (i.e., `_isPower(of:)`).

The conclusion is the fast-/slow-path pattern is a good fit for implementing this API, given that it happens much more often to check if a number is a power of 2 than other bases, and querying for a power of 10 can sometimes be common as well.

jrose-apple commented on stdlib/public/core/Integers.swift:
```diff
@@ -1300,6 +1315,94 @@ extension BinaryInteger {
+        ((exponent << 2) | 0b01) & 0b1111 else { return false }
+
+      // Now time for the slow path.
+      return self._isPower(of: 10)
```
`(self >> exponent)._isPower(of: 5)`?

dingobye replied:
Thank you for the prompt comments, Jordan!

I think this is not a sufficient condition, because it only ensures self == 2^exponent * 5^m, where exponent may not equal m. To make it correct, we'd do something like:
```swift
var powerOfFive = 1
for _ in 0..<exponent { powerOfFive *= 5 } // TODO: handle overflow here
return (self >> exponent) == powerOfFive
```

jrose-apple replied:
Ugh, right, sorry, I jumped ahead to that in my head and then decided it wasn't necessarily worth it.

jrose-apple commented on stdlib/public/core/Integers.swift:
```diff
@@ -1216,6 +1216,21 @@ public protocol BinaryInteger :
+  /// _-27 = -3 * -3 * -3_. *1* is a power of everything because any number to
+  /// the zero power is one.
+  ///
+  /// For the coner case where base is zero, `x.isPower(of: 0)` is `true` if
```
Nitpick: corner

dingobye replied:
Done!

michelf commented:
No optimization for bases that are a power of 2?
https://forums.swift.org/t/adding-ispowerof2-to-binaryinteger/24087/31

dingobye commented:
Thanks for pointing this out, Michel~ I have put this together and will update the performance results later.

jrose-apple commented on stdlib/public/core/Integers.swift:
```diff
@@ -1329,6 +1330,13 @@ extension BinaryInteger {
     return self > 0 && self & (self - 1) == 0
   }
 
+  @usableFromInline @_transparent
```
I don't know if you actually want all these to be `transparent` rather than `inlinable`. Good for constants, potentially bad for non-constants.

dingobye replied:
Yes, it should be `inlinable` here. Thank you for pointing this out!

dingobye commented:
Following @stephentyrone 's advice, I had `_isPowerOfTwo` improved by using the `words`-based implementation. It is now more efficient for those prototype integers (i.e. `DoubleWidth` and `BigInt`).

dingobye commented:
As per the feedbacks from the community, I had the public API `isPower(of:)` downgraded from a protocol requirement to an extension method, thus that it adds less complexity to the `BinaryInteger` protocol's interface.

michelf commented on stdlib/public/core/Integers.swift:
```diff
@@ -1300,6 +1300,152 @@ extension BinaryInteger {
+    // Fast path when base is one of the common cases.
+    if base == 2 { return self._isPowerOfTwo }
+    if base == 10 { return self._isPowerOfTen }
+    if base._isPowerOfTwo { return self._isPowerOf(powerOfTwo: base) }
```
We have a "fast path" for some commonly used bases, but this doesn't necessarily mean the other bases are so infrequent that we should make them slow.

From your link above:

> ... it can actually degrade performance in non-intuitive ways as it marks all other code (including subsequent _fastPaths) as being cold. Consider _fastPath as basically spraying the rest of the code with a Mr. Freeze-style ice gun.

dingobye replied:
Thanks for catching those typos, Ben!

I agree with Michel that "slow path" in this context means _relatively inefficient algorithm_ only, but not *cold path*.

This patch is available to be included into other modules (e.g., swift-preview), since it does not access existing internal implementation of standard library.

futurejones commented:
Can one of the admins verify this patch?

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

dingobye commented:
@shahmishal  Thanks for the information, Mishal! I had the base branch changed to `main`.

@jckarter Hi Joe, I rebased this PR; and added some minor comment, as you suggested in the evolution review, for better clarification purpose. Would you take a look at it and fire a test, please?

stephentyrone commented:
@swift-ci test

stephentyrone commented:
@swift-ci Please test source compatibility

stephentyrone commented on stdlib/public/core/Integers.swift:
```diff
@@ -1315,6 +1315,152 @@ extension BinaryInteger {
+    if base.magnitude <= 1 { return self == base }
+
+    // At this point, we have base.magnitude >= 2. Repeatedly perform
+    // multiplication by a factor of base, and check if it can equal self.
```
The comment here doesn't match the implementation. Please update.

dingobye replied:
Done!

benrimmington commented on stdlib/public/core/Integers.swift:
```diff
@@ -1315,6 +1315,152 @@ extension BinaryInteger {
+  /// Returns `true` if this value is a power of the given base, and `false`
+  /// otherwise.
+  ///
+  /// For two integers *a* and *b*, *a* is a power of *b* if *a* is equal to
+  /// any repeated multiplication of *b*. For example, *16* is a power of *2*
+  /// because _16 = 2 * 2 * 2 * 2_. *-27* is a power of *-3* because
+  /// _-27 = -3 * -3 * -3_. *1* is a power of everything because any number to
+  /// the zero power is one.
```
The examples might be more readable in a list, using code spans instead of emphasis.

```suggestion
  /// For two integers *a* and *b*, *a* is a power of *b* if *a* is equal to
  /// any repeated multiplication of *b*. For example:
  ///
  /// * `16.isPower(of: 2)` because `16 == (2 * 2 * 2 * 2)`
  /// * `(-27).isPower(of: -3)` because `-27 == (-3 * -3 * -3)`
  ///
  /// One is a power of everything because any number to the zero power is one.
```

stephentyrone replied:
> /// One is a power of everything because any number to the zero power is one.

This is not exactly right; one is a power of everything because the empty product is one because one is the multiplicative identity. "Any number to the zero power is one" also follows from the convention for the empty product. I'm not sure how far down this rabbit hole the comment actually needs to go, however =)

dingobye replied:
Yes, this is more precise from a scientific perspective. I have improved the comments by adding `empty product`.

dingobye commented:
Thank Stephen and Ben for your inputs! I had the comments updated.

benrimmington commented on test/Prototypes/BigInt.swift:
```diff
@@ -1871,10 +1871,24 @@ BigIntBitTests.test("words") {
   expectFalse((1 as _BigInt<UInt>).isMultiple(of: 0))
 }
+
+BigIntTests.test("isPower") {
+  expectTrue((0 as _BigInt<UInt>).isPower(of: 0))
+  expectTrue((1 as _BigInt<UInt>).isPower(of: 0))
+  expectTrue((1 as _BigInt<UInt>).isPower(of: 1))
+  expectTrue((1 as _BigInt<UInt>).isPower(of: -1))
+  expectTrue((-1 as _BigInt<UInt>).isPower(of: -1))
+  expectFalse((-1 as _BigInt<UInt>).isPower(of: 1))
+  expectTrue((4 as _BigInt<UInt>).isPower(of: 2))
+  expectTrue((-8 as _BigInt<UInt>).isPower(of: -2))
+  expectTrue((16 as _BigInt<UInt>).isPower(of: -2))
+  expectFalse((32 as _BigInt<UInt>).isPower(of: -2))
+  expectTrue((100 as _BigInt<UInt>).isPower(of: 10))
+}
```
The `BigInt` prototype has been `XFAIL`ed recently. Should this test be removed?

dingobye replied:
How about we keep this test here, in case `BigInt` would be un`XFAIL`ed some time in future? Anyway, I don't mind removing it.

benrimmington replied:
It's not for me to decide, but I'd guess that's OK.

I was going to suggest testing more multi-word values, but DoubleWidth is still available for that purpose.

dingobye replied:
Well, I reverted changes in BigInt, since DoubleWidth, as you mentioned, serves multi-words purpose already. We'd better not touch `XFAIL`'d tests unless necessary.

benrimmington commented on test/api-digester/stability-stdlib-abi-with-asserts.test:
```diff
@@ -46,3 +46,8 @@ Struct _ObjectRuntimeFunctionCountersState is a new API without @available attri
 Struct _RuntimeFunctionCounters is a new API without @available attribute
 Func _swift_isImmutableCOWBuffer(_:) is a new API without @available attribute
 Func _swift_setImmutableCOWBuffer(_:_:) is a new API without @available attribute
+Func BinaryInteger.isPower(of:) is a new API without @available attribute
+Func BinaryInteger._slowIsPower(of:) is a new API without @available attribute
+Func BinaryInteger._isPowerOf(powerOfTwo:) is a new API without @available attribute
+Var BinaryInteger._isPowerOfTen is a new API without @available attribute
+Var BinaryInteger._isPowerOfTwo is a new API without @available attribute
```
These should normally be added to the "…-without-asserts.test" file instead, so that they also apply when building with the `--no-assertions` flag. However, your new APIs do need either `@_alwaysEmitIntoClient` for back-deployment, or `@available` with version 9999 placeholders.

dingobye replied:
Thanks a lot for your information, Ben! I should have read `docs/StandardLibraryProgrammersManual.md` earlier, where the 9999 `@available` placeholder is presented. Now the patch is updated to revert changed in api-digest, and use non-ABI-breaking attributes to those new APIs.

benrimmington replied:
I think the public `isPower(of:)` function can also use `@_alwaysEmitIntoClient` if you want. Then the `#available` guards aren't needed.

dingobye replied:
Done~

benrimmington commented on test/stdlib/Integers.swift.gyb:
```diff
@@ -825,6 +825,35 @@ tests.test("isMultiple") {
 }
 
+tests.test("isPower") {
+  func isPowerTest<T: FixedWidthInteger>(type: T.Type) {
+    expectTrue((0 as T).isPower(of: 0))
+    expectTrue((1 as T).isPower(of: 0))
+    expectTrue((100 as T).isPower(of: 10))
+    expectFalse(T.max.isPower(of: 2))
+    expectFalse(T.min.isPower(of: 2))
+    if T.isSigned {
+      expectTrue((1 as T).isPower(of: -1))
+      expectTrue((-1 as T).isPower(of: -1))
+      expectFalse((-1 as T).isPower(of: 1))
+      expectTrue(T.min.isPower(of: -2))
+      expectTrue(((1 as T) << (T.bitWidth - 2)).isPower(of: 2))
+    } else {
+      expectTrue(((1 as T) << (T.bitWidth - 1)).isPower(of: 2))
+    }
+  }
```
Should this test more values? For example:

```swift
  for value in T(clamping: Int16.min) ... T(clamping: Int16.max) {
    let expected = (value > 0) && (value.nonzeroBitCount == 1)
    let actual = value.isPower(of: 2)
    expectEqual(expected, actual, "(\(value)).isPower(of: 2)")
  }
```

(The copyright year at the top of this file could be updated.)

dingobye replied:
Yes, this is great in coverage. However, checking all 16-bit possibilities seems to be a big thing at runtime, considering this is executed for a number of different `T`s. How about narrowing it from -16 to 16?

benrimmington replied:
Testing all `Int16` values seemed fast enough when I tried it. But maybe use a range which tests all possible `Int8` and `UInt8` values at least?

Can more powers of 10 also be tested in a similar way? And some more tests for each of the slow paths?

(Not only to test your current implementation, but also to test any future changes.)

dingobye replied:
I updated the test to include more comprehensive cases. It is able to test ranges of integers with configurable base values, which is good for testing future changes.

benrimmington commented:
@swift-ci Please test

benrimmington commented:
@swift-ci Please test source compatibility

benrimmington commented:
All checks have passed ([macOS][], [Linux][], [Debug][], [Release][]).

[macOS]:   <https://ci.swift.org/job/swift-PR-osx/23917/>
[Linux]:   <https://ci.swift.org/job/swift-PR-Linux/23933/>
[Debug]:   <https://ci.swift.org/job/swift-PR-source-compat-suite-debug/3215/>
[Release]: <https://ci.swift.org/job/swift-PR-source-compat-suite/4952/>

dingobye commented:
Thank you for your help, Ben! I had the PR rebased with commits squashed.

benrimmington commented on stdlib/public/core/Integers.swift:
```diff
@@ -1315,6 +1315,161 @@ extension BinaryInteger {
+    // Fast path when base is one of the common cases.
+    if base == 2 { return self._isPowerOfTwo }
+    if base == 10 { return self._isPowerOfTen }
+    if base._isPowerOfTwo { return self._isPowerOf(powerOfTwo: base) }
```
Should the `_isPowerOf(powerOfTwo:)` function be removed?

```suggestion
    if base._isPowerOfTwo {
      return self._isPowerOfTwo &&
        self.trailingZeroBitCount.isMultiple(of: base.trailingZeroBitCount)
    }
```

dingobye replied:
I don't think it necessary, though it is good to reduce the number of APIs. IMO, inlining the code seems to make the public function's implementation less concise and less consistent across branches (calls vs inlining code).

Anyway, I think either way is fine. You call.

benrimmington commented:
@swift-ci Please smoke test

mdznr commented on stdlib/public/core/Integers.swift:
```diff
@@ -1315,6 +1315,161 @@ extension BinaryInteger {
+    let (bound, remainder) = self.quotientAndRemainder(dividingBy: base)
+    guard remainder == 0 else { return false }
+
+    // Return true if the product eventualy hits bound. Because if bound is
```
```suggestion
    // Return true if the product eventually hits bound. Because if bound is
```

dingobye replied:
Nice catch, Matt~

stephentyrone commented:
So basically this looks fine to me, except for a weird process detail: during the time period between when this proposal was written and when the SE approval actually ran, Swift Preview was introduced and the process changed, meaning that this feature should actually land first in [Swift Preview](https://github.com/apple/swift-standard-library-preview) rather than directly in the standard library.

It's annoying because while Swift Preview is supposed to decrease the workload for proposals, it increases it for things that happen to have been inflight while the process moved under them. My apologies for that. Please reach out to me and @natecook1000 if you need any assistance in creating a PR against preview.

dingobye commented:
@stephentyrone Thanks for pulling this work on the right track, Stephen!

I created a PR against the `staging` repo [here](https://github.com/apple/swift-evolution-staging/pull/11), with most of copy&paste work.

As for the `preview` repo, looks like we need to have such feature setting up its own repository before re-exporting it from the umbrella library. Is there anything I should do to proceed with the `preview` repo?

stephentyrone commented:
@natecook1000 Can you clarify the process for @dingobye?

stephentyrone commented:
@dingobye There's one other thing we should resolve (which I should have flagged in the proposal review, but missed). Instead of the existing signature:
```
public func isPower(of base: Self) -> Bool
```
this method should be:
```
public func isPower<Base: BinaryInteger>(of base: Base) -> Bool
```

This is a fairly trivial change, but will allow us to avoid creating a bignum object in the case where Self is an arbitrary-precision (or even just large, e.g. Int128 or bigger) integer and base is small. Formally, this requires another evolution proposal, but it can be quite perfunctory, and I'm happy to help write the proposal. I think you should go ahead and adopt this change for your preview package, and we'll get review started as soon as we can.

dingobye commented:
@stephentyrone Hi Stephen, I updated the PR against the staging repo [here](https://github.com/apple/swift-evolution-staging/pull/11/commits/36cd475a7ff08c78d0919936634ff3227d69a27c) as per your advice.

In fact, I took some time thinking about the function signature at the very beginning when pitching this idea. After browsing the existing functions in standard library, I made the decision to use `Self` for consistency consideration. Because standard library is mostly using `Self` for interface functions, with some examples below:
```
func quotientAndRemainder(dividingBy: Self) -> (quotient: Self, remainder: Self)
func isMultiple(of: Self) -> Bool
static func &<<= (inout Self, Self)
static func random(in: ClosedRange<Self>) -> Self
```

Pros of using `base: Self`:
* consistent with most other functions
* users (at least myself) may feel more natural to see it `Self`

Cons of using `base: Self`:
* non-optimal performance when Self is heavy and base is in some lightweight type. However, such kind of limitation also exists in many other standard library functions (e.g., the above shifting operation when shifting 1 bit on a QuadrupleWidth). Is it tolerable?

I don't mind going through review process, since the public function signature is so important to ABI. Not sure if it is really worth the change. Would like to hear more of your inputs.

stephentyrone commented:
> However, such kind of limitation also exists in many other standard library functions (e.g., the above shifting operation when shifting 1 bit on a QuadrupleWidth).

Well, no, because we _also_ have
```
static func &<<= <Other: BinaryInteger>(inout Self, Other) -> Self
```

`isMultiple` should receive the same treatment as `isPower`, and that can be folded into the proposal (we'll have to keep the old version around for binary compatibility reasons, but we can still add the generic one).

random(in:) and quotientAndRemainder(dividingBy:) have semantic constraints that make it so that having them be generic makes a lot less sense, so those are fine as is.

dingobye commented:
@stephentyrone 

> Well, no, because we also have
> static func &<<= <Other: BinaryInteger>(inout Self, Other) -> Self

If we check its implementation, it turns rhs to `Self` via `Self(truncatingIfNeeded: rhs)`. So It _does_ create a QuadrupleWidth for 1, when shifting 1 bit on a QuadrupleWidth. It is non-optimal for lightweight `Other` type, which is the same as this PR. Anyway, having such function signature preserves the possibility/potential for future optimizations without much concern about impacts on ABI.

Can we revise SE-0288 for a 2nd round review, without having to prepare a separate proposal? If it works, would you help revise the proposal and [staging repo's implementation](https://github.com/apple/swift-evolution-staging/pull/11/) to make it armed for a 2nd round review? BTW, don't forget to remove the first acknowledgement item when adding your name into the author list. I've already owned you too much on this story, and you are well deserved. Thank you~

tbkka commented:
Is this PR still live?   Or has it been supplanted by a PR to the Preview repo?

shgew commented on stdlib/public/core/Integers.swift:
```diff
@@ -1315,6 +1315,161 @@ extension BinaryInteger {
+    }
+  }
+
+  /// Returns `true` iff `self` is a power of the given `base`.
```
🔨 
```suggestion
  /// Returns `true` if `self` is a power of the given `base`.
```
